include/zephyr/net/net_core.h

```diff
@@ -154,6 +155,42 @@ int net_send_data(struct net_pkt *pkt);
 #define NET_TC_COUNT 0
 #endif /* CONFIG_NET_TC_TX_COUNT && CONFIG_NET_TC_RX_COUNT */
 
+/**
+ * @brief Registration information for a given L3 handler. Note that
+ *        the layer number (L3) just refers to something that is on top
+ *        of L2. So for example IPv6 is L3 and IPv4 is L3, but Ethernet
+ *        based LLDP, gPTP are more in the layer 2.5 but we consider them
+ *        as L3 here for simplicity.
+ */
+struct net_l3_register {
+	/** Store also the name of the L3 type in order to be able to
+	 * print it later.
+	 */
+	const char * const name;
+	/** Handler function for the given protocol type */
+	enum net_verdict (*handler)(struct net_if *iface,
+				    uint16_t ptype,
+				    struct net_pkt *pkt);
```

Comment:
We can also pass the source and destination addresses here, if the handler wants to know these.

Reply:
Wouldn't those be already set in net_pkt structure (`lladdr_src/dst` fields)?

Reply:
Yes, I have used the values stored in net_pkt